File: plugins/codexy/mcp/codegraph/files.js
"use strict";

const fs = require("fs");
const path = require("path");
const { spawnSync } = require("child_process");

const codeExtensions = new Set([".js", ".jsx", ".ts", ".tsx", ".mjs", ".cjs", ".py", ".go", ".rs", ".rb", ".java", ".kt"]);

function repoRoot(inputRoot) {
  const candidate = inputRoot ? path.resolve(inputRoot) : process.cwd();
  return fs.existsSync(candidate) ? candidate : process.cwd();
}

function resultLimit(inputLimit) {
  const parsed = Number(inputLimit);
  return Number.isFinite(parsed) && parsed > 0 ? Math.floor(parsed) : 80;
}

function includeLine(line) {
  return Boolean(line);
}

function isCodeFile(file) {
  return codeExtensions.has(path.extname(file));
}

function toPosix(file) {
  return file.split(path.sep).join("/");
}

function ignoredByGit(root, relativePaths) {
  if (relativePaths.length === 0) return new Set();
  const input = `${relativePaths.join("\0")}\0`;
  const result = spawnSync("git", ["-C", root, "check-ignore", "-z", "--stdin"], {
    input,
    encoding: "utf8",
    maxBuffer: 1024 * 1024,
  });
  if (result.status === 1) return new Set();
  if (result.status !== 0) return new Set();
  return new Set(result.stdout.split("\0").filter(Boolean));
}

function walkCodeFiles(root) {
  const files = [];
  const ignored = new Set([".git", "node_modules"]);

  function walk(dir) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    const relativeEntries = entries.map((entry) => toPosix(path.relative(root, path.join(dir, entry.name))));
    const gitIgnored = ignoredByGit(root, relativeEntries);

    for (const [index, entry] of entries.entries()) {
      const relative = relativeEntries[index];
      if (ignored.has(entry.name) || gitIgnored.has(relative)) continue;
      const absolute = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(absolute);
        continue;
      }
      if (entry.isFile() && isCodeFile(absolute)) {
        files.push(relative);
      }
    }
  }

  walk(root);
  return files.sort();
}

function unique(values) {
  return Array.from(new Set(values)).sort();
}

module.exports = {
  codeExtensions,
  includeLine,
  isCodeFile,
  ignoredByGit,
  repoRoot,
  resultLimit,
  toPosix,
  unique,
  walkCodeFiles,
};

File: src/codegraph/files.rs
use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf};

use ignore::WalkBuilder;

const CODE_EXTENSIONS: &[&str] = &[
    "js", "jsx", "ts", "tsx", "mjs", "cjs", "py", "go", "rs", "rb", "java", "kt", "html", "htm",
    "css", "scss", "sass", "less", "svg", "vue", "svelte", "astro", "json", "jsonc", "yaml", "yml",
    "toml", "md", "mdx",
];

pub(super) fn result_limit(input: Option<usize>) -> usize {
    input.filter(|value| *value > 0).unwrap_or(80)
}

pub(super) fn repo_root(input_root: Option<&str>) -> PathBuf {
    let current_dir = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    let candidate = input_root.map_or_else(|| current_dir.clone(), PathBuf::from);
    let rooted = if candidate.is_absolute() {
        candidate
    } else {
        current_dir.join(candidate)
    };
    if rooted.exists() {
        rooted.canonicalize().unwrap_or(rooted)
    } else {
        current_dir
    }
}

pub(super) fn is_code_file(path: &Path) -> bool {
    path.extension()
        .and_then(|extension| extension.to_str())
        .is_some_and(|extension| CODE_EXTENSIONS.contains(&extension))
}

pub(super) fn code_extensions() -> BTreeSet<String> {
    CODE_EXTENSIONS
        .iter()
        .map(|extension| format!(".{extension}"))
        .collect()
}

pub(super) fn to_posix(path: &Path) -> String {
    path.components()
        .map(|component| component.as_os_str().to_string_lossy())
        .collect::<Vec<_>>()
        .join("/")
}

pub(super) fn walk_code_files(root: &Path) -> Vec<String> {
    let mut files = WalkBuilder::new(root)
        .hidden(false)
        .filter_entry(|entry| {
            let name = entry.file_name().to_string_lossy();
            name != ".git" && name != "node_modules"
        })
        .build()
        .filter_map(Result::ok)
        .filter(|entry| {
            entry
                .file_type()
                .is_some_and(|file_type| file_type.is_file())
        })
        .filter(|entry| is_code_file(entry.path()))
        .filter_map(|entry| entry.path().strip_prefix(root).ok().map(to_posix))
        .collect::<Vec<_>>();
    files.sort();
    files
}

pub(super) fn read_source(root: &Path, file: &str) -> String {
    fs::read_to_string(path_join_posix(root, file)).unwrap_or_default()
}

fn path_join_posix(root: &Path, file: &str) -> PathBuf {
    file.split('/')
        .fold(root.to_path_buf(), |path, part| path.join(part))
}
